### PR TITLE
Create plugin architecture

### PR DESCRIPTION
**Pull request by dstallenberg**
This PR implements the new plugin architecture that allows splitting unused parts of the code into plugins and connecting uncommonly used code to the code.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -33,17 +33,18 @@
...
+    "format:check": "prettier --config .prettierrc.json --ignore-path .prettierignore --check .",
+    "lint": "eslint --ignore-path .eslintignore .",
+    "lint:fix": "eslint --ignore-path .eslintignore . --fix",
     "prepare": "npm run build",
-    "test": "mocha",
-    "test:coverage": "nyc mocha",
-    "test:watch": "mocha --watch"
+    "test": "mocha --config .mocharc.json",
+    "test:coverage": "nyc mocha --config .mocharc.json",
+    "test:watch": "mocha --config .mocharc.json --watch"
```
Comment (mitchellolsthoorn) [inline package.json:43]: These changes to the NPM run scripts should probably be in a separate PR. This also helps document why this needs to be changed

**Comment by dstallenberg** [comment]
This PR is superseded by #125 as it was impossible to implement the plugin architecture without the event manager as they are tightly integrated.
